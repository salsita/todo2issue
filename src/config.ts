import { config as loadDotenv } from 'dotenv'
import { resolve } from 'path'
import { readFile } from 'fs/promises'
import { promisify } from 'util'
import readJson from 'read-package-json'
import gitUrlParse from 'git-url-parse'

const readJsonAsync = promisify(readJson)

loadDotenv({
  path: resolve(process.cwd(), '.env')
})

export const GITHUB_TOKEN = process.env.GITHUB_TOKEN

export interface GitRepository {
  name: string
  owner: string
}

export interface Config {
  filePatterns: string[]
  repo: GitRepository
}

const defaultConfig: Partial<Config> = {
  filePatterns: ['**/*.[tj]s?(x)']
}

export async function readConfig (root: string): Promise<Config> {
  const {
    repository,
    todo2issue: {
      filePatterns = defaultConfig.filePatterns
    } = defaultConfig
  } = await readJsonAsync(resolve(root, 'package.json'))

  if (!repository?.url) {
    throw new Error(`'repository.url' missing in 'package.json'`)
  }

  const { name, owner } = gitUrlParse(repository?.url)

  if (!Array.isArray(filePatterns) || filePatterns.some(pattern => typeof pattern !== 'string')) {
    throw new Error(`'todo2issue.filePatterns' property is expected to be an array of glob patterns`)
  }
  return {
    filePatterns,
    repo: { name, owner }
  }
}
