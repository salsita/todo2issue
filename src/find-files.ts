import globby from 'globby'

export function findFiles(root: string, patterns: string[]): Promise<string[]> {
  return globby(patterns, {
    cwd: root,
    expandDirectories: false,
    gitignore: true
  })
}
