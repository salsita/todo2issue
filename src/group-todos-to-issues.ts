import { Issue, Todo } from './model'

export function groupTodosToIssues(todos: Todo[]): Issue[] {
  const issues: Issue[] = []
  const issuesByNumber = new Map<number, Issue>()
  todos
    .filter((todo) => !!todo.issueNumber)
    .forEach((todo) => {
      let issue = issuesByNumber.get(todo.issueNumber!)
      if (!issue) {
        issue = { todos: [], issueNumber: todo.issueNumber }
        issuesByNumber.set(todo.issueNumber!, issue)
        issues.push(issue)
      }

      issue.todos.push(todo)
    })

  const issuesByText = new Map<string, Issue>()

  Array.from(issuesByNumber.values()).forEach((issue) => {
    issue.todos.forEach((todo) => issuesByText.set(todo.text, issue))
  })

  todos
    .filter((todo) => !todo.issueNumber)
    .forEach((todo) => {
      let issue = issuesByText.get(todo.text)
      if (!issue) {
        issue = { todos: [] }
        issuesByText.set(todo.text, issue)
        issues.push(issue)
      }
      issue.todos.push(todo)
    })

  return issues
}
