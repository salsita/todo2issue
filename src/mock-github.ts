import { GithubClient, PartialGithubIssue } from './github'
import { Issue } from './model'

const mockIssues: PartialGithubIssue[] = [
  {issueNumber: 123, body: 'Some'},
  {issueNumber: 124, body: 'Somewhat else'},
  {issueNumber: 125, body: 'Completely else'},
  {issueNumber: 126, body: 'Different league'},
]

export class MockGithubClient implements GithubClient {
  protected issueNumberCounter: number | null = null
  public readonly log: string[] = []

  async createIssue (issue: Issue, issueLabel: string, body: string): Promise<number> {
    if(this.issueNumberCounter === null) {
      this.issueNumberCounter = 0
    }
    const issueNumber = ++this.issueNumberCounter
    this.log.push(`created issue #${issueNumber} for\n${JSON.stringify(issue, null, 2)}`)
    return issueNumber
  }

  async updateIssue (issueNumber: number, body: string): Promise<void> {
    this.log.push(`updated issue #${issueNumber}`)
  }

  async listOpenTodoIssues (issueLabel: string): Promise<PartialGithubIssue[]> {
    this.log.push(`listed issues ${mockIssues.map(issue => `#${issue.issueNumber}`).join(', ')}`)
    return mockIssues
  }

  async closeIssue (issueNumber: number) {
    this.log.push(`close issues #${issueNumber}`)
  }
}

export class WriteMockGithubClient extends MockGithubClient {
  constructor (
    private realClient: GithubClient,
    sequenceStart?: number,
    private readonly sequenceEnd?: number
  ) {
    super()
    if (sequenceStart) {
      this.issueNumberCounter = sequenceStart - 1
    }
  }

  async createIssue (issue: Issue, issueLabel: string, commitish: string): Promise<number> {
    if(this.issueNumberCounter === null) {
      const existingIssues = await this.realClient.listOpenTodoIssues(issueLabel)
      this.issueNumberCounter = Math.max(0, ...existingIssues.map(issue => issue.issueNumber))
    }
    if (this.issueNumberCounter >= this.sequenceEnd) {
      throw new Error('sequence exhausted')
    }
    return super.createIssue(issue, issueLabel, commitish);
  }

  async listOpenTodoIssues (issueLabel: string): Promise<PartialGithubIssue[]> {
    return this.realClient.listOpenTodoIssues(issueLabel);
  }
}
