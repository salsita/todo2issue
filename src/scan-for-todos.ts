import { readFile } from 'fs/promises'
import { extname, resolve } from 'path'
import { parse } from 'leasot'
import { Todo } from './model'

const ISSUE_NUMBER_PATTERN = /^#([0-9]+)$/

export async function scanForTodos(root: string, files: string[]): Promise<Todo[]> {
  const todos: Todo[] = []
  for (const filename of files) {
    const contents = await readFile(resolve(root, filename), 'utf8')
    const todoComments = parse(contents, { filename, extension: extname(filename) })

    todos.push(
      ...todoComments.map((todoComment) => {
        const { line, text, tag, ref } = todoComment
        const [, issueNumber] = ISSUE_NUMBER_PATTERN.exec(ref) || []
        const todo = {
          filename,
          line,
          tag,
          text,
          issueNumber: issueNumber ? Number(issueNumber) : undefined,
        }
        return todo
      }),
    )
  }
  return todos
}
