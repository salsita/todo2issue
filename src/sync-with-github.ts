import { Issue } from './model'
import { GithubClient } from './github'

export async function syncWithGitHub (issues: Issue[], githubClient: GithubClient) {
  for (const issue of issues) {
    if (issue.issueNumber === undefined) {
      issue.issueNumber = await githubClient.createIssue(issue)
    }
  }
  const githubIssues = new Set(await githubClient.listOpenTodoIssueNumbers())
  issues.forEach(issue => githubIssues.delete(issue.issueNumber))

  for (const obsoleteIssueNumber of githubIssues) {
    await githubClient.closeIssue(obsoleteIssueNumber)
  }
}
