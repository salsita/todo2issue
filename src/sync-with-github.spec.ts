import { findFiles } from './find-files'
import { resolve } from 'path'
import { scanForTodos } from './scan-for-todos'
import { groupTodosToIssues } from './group-todos-to-issues'
import { syncWithGitHub } from './sync-with-github'
import { MockGithubClient } from './mock-github'
import { readConfig } from './config'

describe('syncWithGithub', () => {
  it('syncs', async () => {
    const root = resolve(__dirname, '../')
    const config = await readConfig(root)
    const files = await findFiles(root, config.filePatterns)
    const todos = await scanForTodos(root, files)
    const issues = groupTodosToIssues(todos)
    const mockGithubClient = new MockGithubClient([
      { issueNumber: 123, body: 'Some' },
      { issueNumber: 124, body: 'Somewhat else' },
      { issueNumber: 125, body: 'Completely else' },
      { issueNumber: 126, body: 'Different league' }
    ])
    await syncWithGitHub(issues, mockGithubClient, config.repo, 'TODO', 'master', false)
    expect(mockGithubClient.log).toMatchSnapshot()
  })
})
