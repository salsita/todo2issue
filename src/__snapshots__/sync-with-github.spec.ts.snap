// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syncWithGithub syncs 1`] = `
Array [
  "created issue #1 for Single line todo comment",
  "created issue #2 for line comments",
  "created issue #3 for line comments with reference",
  "created issue #4 for With colon",
  "created issue #5 for This one is from a sub-folder",
  "listed issues #123, #124, #125, #126",
  "close issues #126",
]
`;
